Update Form's Minimum and Maximum sizes with DPI_CHANGED event (#7467)
dotnet/winforms: docs/WinformsRuntimeConfiguration.md

> **Runtime configuration in legacy .NET framework applications**

Author: Lets chat offlien on this.

generators at compile time. This document outlines expansion of those application wide configurations further to cover runtime config options.

**runtimeconfig.json for Winforms .NET applications.**

app.config has limited support in .NET and goal is to move away from using it in .NET for
performance and reliability reasons. .NET runtime and other .NET teams use runtimeconfig.json to define .NET runtime configurations and appsettings.json to define application-level settings. In this proposal, we are leveraging runtimeconfig.json to define WinForms runtime configurations.

While this proposal is focusing on providing an alternative solution for existing configuration sections `AppContextSwitchOverrides` and `System.Windows.Forms.ApplicationConfigurationSection` that are primarily used for specifying feature flags impacting winforms runtime behavior, we will be looking into alternatives for `Application Settings` that doesn't require app.config in the upcoming releases of .NET.

**Goals:**

> -   Replacement for `AppContextSwitchOverrides` and `System.Windows.Forms.ApplicationConfigurationSection` of app.config.
> 
> -   Users should be able to update/modify Winforms applications runtime configurations without recompiling the application
> 
> -   Existing applications should be able to seamlessly upgrade to this
>     new model when targeting to latest .NET.
> -   Existing [build time properties]((https://docs.microsoft.com/en-us/dotnet/desktop/winforms/whats-new/net60?view=netdesktop-6.0#new-application-bootstrap)) defined via project file continue to work.

Reviewer: ```suggestion
-   A replacement for `AppContextSwitchOverrides` and `System.Windows.Forms.ApplicationConfigurationSection` of app.config.
-   Users should be able to update/modify Winforms applications runtime configurations without recompiling the application.
-   The existing applications should be able to seamlessly upgrade to the new configuration model when targeting the latest .NET runtiem.
-   The existing [application configuration MSBuild properties](https://aka.ms/applicationconfiguration) continue to work.
```

**Out of scope:**

> -  App settings that are serialized from Settings designer/editor page. Applications should continue to use current model until we come up with an alternative solution for this.
> -  Dynamic/real-time loading of configuration values from runtimeconfig.json.
> -  Unification of Build/Runtime configurations into one place, runtimeconfig.json.
>     -   We keep current build time configurations in the project file while adding new runtime configurations into runtimeconfig.json. We may revisit build time properties for unification in the upcoming releases of .NET.

Reviewer: ```suggestion
-  App settings that are serialized by the Settings designer/editor page. Applications should continue to use the current model.
-  Dynamic/real-time (re-)loading of configuration values from runtimeconfig.json.
-  Unification of comple/runtime configurations into one place, runtimeconfig.json.
```

The last item is redundant, it's called out as the last option under the Goals section.

**Syntax of runtimeConfig.template.Json**.

```xml
{
  "configProperties": {
      "System.Globalization.UseNls": true,
      "System.Net.DisableIPv6": true,
      "System.GC.Concurrent": false,
      "System.Threading.ThreadPool.MinThreads": 4,
      "System.Threading.ThreadPool.MaxThreads": 25
    }
}
```

> Given this section is [common across .NET products](https://docs.microsoft.com/en-us/dotnet/core/runtime-config), Winforms specific switches will have unique names for its feature flags to avoid conflicts with other .NET teams. Following is the syntax that winforms adapt while defining a new feature switch/flag.
> 
>    Winforms runtime flags/settings: "System.Windows.Forms.<runtime_flag/setting_name>"
> 
> ex: 
> ```xml
> {
>   "configProperties": {
>       "System.Globalization.UseNls": true,
>       "System.Net.DisableIPv6": true,
>       "System.GC.Concurrent": false,
>       "System.Threading.ThreadPool.MinThreads": 4,
>       "System.Threading.ThreadPool.MaxThreads": 25,
>       "System.Windows.Forms.ScaleTopLevelFormMinMaxSize": true,
>       "System.Windows.Forms.<CustomEnumProperty>": "EnumValue"
>     }
> }
> ```

Reviewer: ````suggestion
Windows Forms switches will be added to [`configProperties` section](https://docs.microsoft.com/dotnet/core/runtime-config) with the rest of the .NET switches. To avoid name conflics Windows Forms specific switches will be prefixed with `System.Windows.Forms`:
```xml
{
  "configProperties": {
      "System.Globalization.UseNls": true,
      "System.Net.DisableIPv6": true,
      "System.GC.Concurrent": false,
      "System.Threading.ThreadPool.MinThreads": 4,
      "System.Threading.ThreadPool.MaxThreads": 25,
      
      <!-- Windows Forms specific switches -->
      "System.Windows.Forms.ScaleTopLevelFormMinMaxSize": true,
      "System.Windows.Forms.<CustomEnumProperty>": "EnumValue"
    }
}
```
````

**Reading Winforms runtime configurations:**

> Build system [generate](https://docs.microsoft.com/en-us/dotnet/core/runtime-config) `<AppName>.runtimeconfig.json` file in the output directory with the content specified in the `runtimeconfig.template.json` file along with application target framework information. .NET runtime recognizes this file and get it loaded while running Winforms applications.
> 
> In this proposal, we will be focusing primarily on enabling only runtime configurations for WinForms applications. Support for analyzers and source generators are considered in later .NET releases. We will revisit this implementation to improve user-experience further, as we make progress.
> 
> Ex: Content of <AppName>.runtimeconfig.json generated from above content.

Reviewer: ```suggestion
When a project is [built](https://docs.microsoft.com/en-us/dotnet/core/runtime-config/#runtimeconfigjson), an _[appname].runtimeconfig.json_ file is generated in the output directory. If a _runtimeconfig.template.json_ file exists in the same folder as the project file, any configuration options it contains are inserted into the _[appname].runtimeconfig.json_ file.

This proposal focuses primarily on enabling the runtime configurations for WinForms applications. A support for analyzers and source generators may be considered in the future .NET releases. We will revisit this implementation to improve user-experience further, as we make progress.

For example, the content of _[appname].runtimeconfig.json generated from above content:
```

> Ex: Content of <AppName>.runtimeconfig.json generated from above content.

Reviewer: The official docs use `[appname].runtimeconfig.json` format, we should use the same for the consistency.

```XML
{
  "runtimeOptions": {
    "tfm": "net7.0",
    "frameworks": [
      {
        "name": "Microsoft.NETCore.App",
        "version": "7.0.0-preview.7.22375.6"
      },
      {
        "name": "Microsoft.WindowsDesktop.App",
        "version": "7.0.0-preview.7.22377.1"
      }
    ],
    "configProperties": {
      "System.Globalization.UseNls": true,
      "System.Net.DisableIPv6": true,
      "System.GC.Concurrent": false,
      "System.Threading.ThreadPool.MinThreads": 4,
      "System.Threading.ThreadPool.MaxThreads": 25,
      "Switch.System.Windows.Forms.ScaleTopLevelFormMinMaxSize": true,
      "System.Windows.Forms.<StringProperty>": "string"
      "System.Windows.Forms.<CustomProperty>": "CustomValue"
    }
  }
}
```

> Target framework information added to <AppName>.runtimeconfig.json file always match with application target framework irrespective of runtime/SDK installed on the machine or runtime used (Roll-forward scenarios) by the WinForms application.

Reviewer: ```suggestion
The target framework information added to <AppName>.runtimeconfig.json file always match with the application target framework irrespective of runtime/SDK installed on the machine or runtime used ([roll-forward scenarios](https://docs.microsoft.com/dotnet/core/versions/selection#framework-dependent-apps-roll-forward)) by the Windows Forms application.
```

**.NET runtime support in reading runtimeconfig.json:**

WinForms leveraging [the support](https://github.com/dotnet/runtime/blob/5098d45cc1bf9649fab5df21f227da4b80daa084/src/native/corehost/runtime_config.cpp) provided by .NET runtime in reading the `<AppName>.runtimeconfig.json` file and thus get all plumbing required for various hosting and environment scenarios. 

> WinForms add following wrapper on top of runtime implementation to improve performance by caching configuration options and also to define custom defaults values for the WinForms specific feature flags. This is very similar to the [implementation](https://github.com/dotnet/runtime/blob/04dac7b0fede29d44f896c5fd793754f83974175/src/libraries/System.Private.CoreLib/src/System/AppContextConfigHelper.cs) used in various .NET runtime components.

Reviewer: I'm not sure how to read this correctly.

`WinForms adds... wrapper... _to improve_ performance by caching... _and also to define_ custom defaults...` - i.e., add a wrapper to improve and to define

- or -

`WinForms adds... wrapper... to improve performance _by caching_... _and also by defining_ custom defaults...` - i.e., add a wrapper to improve by caching and by defining.

Author: intended to be read as `i.e., add a wrapper to improve and to define`

> WinForms leveraging [the support](https://github.com/dotnet/runtime/blob/5098d45cc1bf9649fab5df21f227da4b80daa084/src/native/corehost/runtime_config.cpp) provided by .NET runtime in reading the `<AppName>.runtimeconfig.json` file and thus get all plumbing required for various hosting and environment scenarios. 
> 
> WinForms add following wrapper on top of runtime implementation to improve performance by caching configuration options and also to define custom defaults values for the WinForms specific feature flags. This is very similar to the [implementation](https://github.com/dotnet/runtime/blob/04dac7b0fede29d44f896c5fd793754f83974175/src/libraries/System.Private.CoreLib/src/System/AppContextConfigHelper.cs) used in various .NET runtime components.
> 
> Below example illustrate making the feature flag `ScaleTopLevelFormMinMaxSizeForDpi` default opt-in if application is running on Windows 10 and targeting .NET 8.

Reviewer: ```suggestion
WinForms is leveraging the [support](https://github.com/dotnet/runtime/blob/5098d45cc1bf9649fab5df21f227da4b80daa084/src/native/corehost/runtime_config.cpp) provided by .NET runtime in reading the `<AppName>.runtimeconfig.json` file, and thus get all plumbing required for various hosting and environment scenarios. 

WinForms adds the following wrapper on top of .NET runtime implementation to improve the performance by caching configuration options, and also by defining custom defaults values for the WinForms specific feature flags. This wrapper is very similar to the [.NET runtime's implementation](https://github.com/dotnet/runtime/blob/04dac7b0fede29d44f896c5fd793754f83974175/src/libraries/System.Private.CoreLib/src/System/AppContextConfigHelper.cs).

The below example illustrates how to define `ScaleTopLevelFormMinMaxSizeForDpi` feature flag, opt-in by default, if an application is running on Windows 10 and is targeting .NET 8:
```

```cs
 internal static partial class LocalAppContextSwitches

>     {

Reviewer: Incorrect indentation

        // Property to cache feature flag "Switch.System.Windows.Forms.ScaleTopLevelFormMinMaxSize" value. 
        private static int s_scaleTopLevelFormMinMaxSize;

>         public static bool ScaleTopLevelFormMinMaxSize

Reviewer: Missing blank line

            if (!hasSwitch)
            {
                isSwitchEnabled = GetSwitchDefaultValue(switchName);
            }

            // Is caching switches disabled?.
            AppContext.TryGetSwitch("TestSwitch.LocalAppContext.DisableCaching", out bool disableCaching);
            if (!disableCaching)
            {
                cachedSwitchValue = isSwitchEnabled ? 1 /*true*/ : -1 /*false*/;
            }

            return isSwitchEnabled;
        }

        // Provides default values for switches if they're not always false by default
        private static bool GetSwitchDefaultValue(string switchName)
        {
            if (OsVersion.IsWindows10_1703OrGreater)
            {
                var tfm = RuntimeTargetFramework.Framework;
                if (tfm is not null && tfm.Name == "Microsoft.NETCore.App" && string.Compare(tfm.Version,"8.0", StringComparison.OrdinalIgnoreCase) >=0)
                {
                    if (switchName == "Switch.System.Windows.Forms.ScaleTopLevelFormMinMaxSizeForDpi")
                    {
                        return true;
                    }
                }
            }

>             return false;

Reviewer 2: Would it be a problem if this switch is set to true on OS lower than 1703?

Author: This particular switch is for PermonV2 scenarios, and we do not want it to be defaulted to true when running on older OSes. as far as i know, .NET applications are not blocked running on Win7/Win8.

Reviewer 2: WIn7/8 will not have PerMonitorV2 mode, so I assume the new code will not be executed at all even if the user manually sets this quirk, right?

        }

>     }

Reviewer: The sample code doesn't look valid (did you manually craft it and didn't update all references correctly?). E.g., `ScaleTopLevelFormMinMaxSize` refers to `GetCachedSwitchValue`, but no such method is present.

nit: "Switch.System.Windows.Forms.ScaleTopLevelFormMinMaxSizeForDpi" should be turned in to a const.

The implementation of `GetCachedSwitchValueInternal` also looks somewhat suboptimal - i.e., we read the switch and only then check if we need to return the cached value of not. I think that should be reversed, i.e. - if we need to return the cached value then there's no need to read the switch at all.

Author: right. it was altered later and did not update doc with latest source.

```

> WinForms runtime then use the static `LocalAppContextSwitches` class to access runtime configurations. Below is sample accessing feature switch `Switch.System.Windows.Forms.ScaleTopLevelFormMinMaxSize`. 

Reviewer: ```suggestion
WinForms runtime then uses the static `LocalAppContextSwitches` class to access the runtime configurations. The below sample demonstrates how to access the feature switch `Switch.System.Windows.Forms.ScaleTopLevelFormMinMaxSize`:
```

Ex: Use of feature switch in OnDpiChanged() method in Form.cs to scale Min/Max size of the Form.

```cs

>  protected virtual void OnDpiChanged(DpiChangedEventArgs e)

Reviewer 2: nit: formatting

>         {

Reviewer: Incorrect indentation
